src/ucp/core/ucp_am.c

```diff
@@ -0,0 +1,682 @@
+        if (ucs_likely((ssize_t)length <= ucp_ep_config(ep)->am.max_short)) {
+            status = ucp_am_send_short(ep, id, payload, length);
+            if (ucs_likely(status != UCS_ERR_NO_RESOURCE)) {
+                UCP_EP_STAT_TAG_OP(ep, EAGER);
+                ret = UCS_STATUS_PTR(status);
+                goto out;
+            }
+        }
+    }
+    
+    status = ucp_ep_resolve_dest_ep_ptr(ep, ep->am_lane);
+    if (status != UCS_OK) {
+        ret = UCS_STATUS_PTR(status);
+        goto out;
+    }
+
+    req = ucp_request_get(ep->worker);
+    if (ucs_unlikely(req == NULL)) {
+        ret = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+        goto out;
+    }
+
+    ucp_am_send_req_init(req, ep, payload, datatype, count, flags, id);
+    if (flags & UCP_AM_SEND_REPLY) {
+        ret = ucp_am_send_req(req, count, &ucp_ep_config(ep)->am, cb,
+                              ucp_ep_config(ep)->am_u.reply_proto);
+    } else {
+        ret = ucp_am_send_req(req, count, &ucp_ep_config(ep)->am, cb,
+                              ucp_ep_config(ep)->am_u.proto);
+    }
+
+out:
+    UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(ep->worker);
+    return ret;
+}
+
+
+static ucs_status_t
+ucp_am_handler_common(ucp_worker_h worker, void *hdr_end,
+                      size_t hdr_size, size_t args_length,
+                      ucp_ep_h reply_ep, uint16_t am_id, 
+                      uint16_t desc_flag, unsigned am_flags) {
+    ucp_recv_desc_t *desc   = NULL;
+    uint16_t recv_flags = 0;
+    ucs_status_t status;
+
+    if (ucs_unlikely(am_flags & UCT_CB_PARAM_FLAG_DESC)) {
+        recv_flags |= desc_flag;
+    }
+
+    status = ucp_recv_desc_init(worker, hdr_end, hdr_size + args_length,
+                                0, am_flags, 0,
+                                recv_flags, 0, &desc);
+    status = worker->am_cbs[am_id].cb(worker->am_cbs[am_id].context,
+                                      desc + 1,
+                                      args_length,
+                                      reply_ep,
+                                      UCP_CB_PARAM_FLAG_DATA);
+
+    if ((status == UCS_OK) && !(am_flags & UCT_CB_PARAM_FLAG_DESC)) {
+        ucp_recv_desc_release(desc);
+    } else if ((status == UCS_OK) &&
+               (am_flags & UCT_CB_PARAM_FLAG_DESC)) {
+        status = UCS_OK;
+    } else if (status != UCS_OK && status != UCS_INPROGRESS) {
+        ucs_error("Unsupported return value");
+    } else if (status == UCS_INPROGRESS &&
+               !(am_flags & UCT_CB_PARAM_FLAG_DESC)) {
+        status = UCS_OK;
+    }
+
+    return status;
+}
+
+static ucs_status_t 
+ucp_am_handler_reply(void *am_arg, void *am_data, size_t am_length,
+                     unsigned am_flags)
+{
+
+    ucp_am_reply_hdr_t *hdr = (ucp_am_reply_hdr_t *) am_data;
+    ucp_worker_h worker     = (ucp_worker_h) am_arg;
+    uint16_t am_id          = hdr->super.am_id;
+    ucp_ep_h reply_ep;
+
+    if (ucs_unlikely((hdr->super.am_id > worker->am_cb_array_len) ||
```

Comment:
This comment is not addressed

Reply:
@evgeny-leksikov your PR GitHu approval confused me...

Reply:
Yes, fixed one check, but not both. Squashed the check to the common function now.

Reply:
@shamisp there are a lot of changes since that approval, we are on next round

Reply:
The original approval came before I had to extend the API for replies.

Reply:
@evgeny-leksikov I have no way of seeing when approval was granter. In the future please revoke the approval and change it to "requested changes".  Otherwise It can be merge accidentally under assumption that any pending concerns were addressed.